### PR TITLE
Crypto Position Lifecycle: Close-Position API Migration & Exit Logic Hardening

### LINKED ISSUE #301: Crypto Position Lifecycle: Close-Position API Migration & Exit Logic Hardening
# Crypto Position Lifecycle: Close-Position API Migration & Exit Logic Hardening

## Background

Our system currently manages crypto positions using simple market orders for entry, with all TP/SL logic tracked in Firestore and evaluated per-run against daily candle data. This is an intentional design due to **Alpaca's crypto order constraints** (see Appendix A). With Issues #297 (reconciler race condition) and #295 (observability) being addressed, this issue focuses on two remaining improvements to the crypto position lifecycle.

## Alpaca Crypto Order Constraints (Reference)

Per [Alpaca Crypto Orders docs](https://docs.alpaca.markets/docs/crypto-orders):

| Feature | Supported | Notes |
|---|---|---|
| `market` orders | ✅ | What we use for entries + exits |
| `limit` orders | ✅ | Available, not currently used |
| `stop_limit` orders | ✅ | Only stop-type order for crypto |
| `stop` orders (plain) | ❌ | **Not supported** |
| Bracket / OTOCO | ❌ | Error `42210000` — equity only |
| TIF: `gtc` | ✅ | What we use |
| TIF: `ioc` | ✅ | Available, not used |
| TIF: `day` | ❌ | Not supported for crypto |
| `DELETE /v2/positions/{symbol}` | ✅ | Supports `qty` and `percentage` params |
| Fractional orders | ✅ | What we use |

**Key constraint**: No bracket orders means TP/SL must be software-managed. No plain `stop` orders means the only broker-side stop option is `stop_limit` (which has gap risk).

## Current System Behavior (What Works)

The system correctly handles per-run evaluation through a well-designed state machine:

### Signal Status Tracking (`check_exits` in `signal_generator.py`)
- **WAITING** → evaluates TP1 (`high >= take_profit_1`), TP2, TP3, invalidation
- **TP1_HIT** → evaluates TP2, TP3, trailing stop updates
- **TP2_HIT** → evaluates TP3, trailing stop updates
- **TP3_HIT / INVALIDATED** → terminal states, triggers close

### TP Automation (`main.py:751-887`)
- **TP1_HIT** → `scale_out_position(0.5)` + `move_stop_to_breakeven()` with idempotency guard (`scaled_out_qty > 0`)
- **TP2_HIT** → `scale_out_position(0.5)` of remaining + `modify_stop_loss(tp1_level)` with idempotency guard (`scaled_out_qty > original * 0.6`)
- **TP3_HIT** → `close_position_emergency()` — full exit
- **INVALIDATED** → `close_position_emergency()` — full exit
- **Status Jump** (WAITING → TP2_HIT in one candle) → handled correctly — TP1 block runs for both TP1_HIT and TP2_HIT

### Firestore State Per Position
- `current_stop_loss`: tracked but only used for equity stop order modifications
- `scaled_out_qty` / `scaled_out_price`: idempotency + weighted average PnL
- `breakeven_applied`: prevents duplicate breakeven moves
- `original_qty`: captured on first scale-out for percentage calculations

**Bottom line**: The per-run state evaluation logic is correct and robust.

## Problem Description

Two issues remain:

### Issue 1: `submit_order(SELL)` instead of `DELETE /v2/positions` for exits

Current exit methods (`close_position_emergency`, `scale_out_position`) submit **new market sell orders** via `submit_order(MarketOrderRequest(SELL, qty=X))` instead of using Alpaca's native `DELETE /v2/positions/{symbol}` endpoint.

This creates problems:
- **Qty mismatch risk**: If Firestore qty drifts from Alpaca's actual qty (as demonstrated in Issue #297), the sell order could be for the wrong amount
- **No `percentage` support**: Can't say "close 50%" — must calculate exact qty, which is fragile for fractional crypto
- **Not idempotent**: A retry of `submit_order(SELL)` could double-sell if the first order was actually filled

```python
# src/crypto_signals/engine/execution.py (close_position_emergency, line ~1530)
# ❌ Current: submits a NEW sell order (needs exact qty from Firestore)
close_request = MarketOrderRequest(
    symbol=position.symbol,
    qty=position.qty,  # <-- If stale, wrong amount
    side=close_side,
    time_in_force=TimeInForce.GTC,
)
close_order = self.alpaca.submit_order(close_request)

# ✅ Better: use Alpaca's close_position (supports percentage, uses actual broker qty)
# DELETE /v2/positions/{symbol_or_asset_id}?percentage=100
# or via alpaca-py: self.alpaca.close_position(symbol, qty=qty, percentage=percentage)
```

Same pattern in `scale_out_position()`:
```python
# src/crypto_signals/engine/execution.py (scale_out_position, line ~1260)
# ❌ Current: calculates qty from Firestore, could be stale
scale_qty = position.qty * scale_pct
close_request = MarketOrderRequest(
    symbol=position.symbol,
    qty=scale_qty,
    side=close_side,
    time_in_force=TimeInForce.GTC,
)

# ✅ Better: use DELETE /v2/positions/{symbol}?percentage=50
# Alpaca calculates qty from its own source of truth
```

### Issue 2: `current_stop_loss` is tracked but never evaluated as exit trigger

The Position's `current_stop_loss` field is written to Firestore and updated by `modify_stop_loss()` and trailing stop syncs. However, `check_exits()` never reads this field — it evaluates SL conditions using:
- **Chandelier Exit** from daily indicator (`current_close < chandelier_exit_long`)  
- **Signal's `invalidation_price`** (static structural level)

This means the stop-loss progression (entry SL → breakeven → TP1 → trailing) is tracked in Firestore but **has no effect on the actual exit decision for crypto**. The exit is always driven by candlestick indicators, not the tracked stop.

```python
# src/crypto_signals/engine/execution.py (modify_stop_loss, line 1140)
# We update current_stop_loss in Firestore...
position.current_stop_loss = new_stop  # ← Persisted to Firestore

# ...but check_exits() never reads it. It uses:
# signal_generator.py:748-756
if is_long and current_close < chandelier_exit:  # ← From daily indicator
    tp3_exit_triggered = True  # current_stop_loss is never consulted
```

This is not necessarily a bug — the Chandelier Exit is a more dynamic indicator. But it means:
- The `current_stop_loss` field is decorative for crypto
- The breakeven move and TP1 stop move are logged but have no protective effect
- If Chandelier Exit diverges from `current_stop_loss`, the tracked stop is meaningless

### 1. Migrate exits to `DELETE /v2/positions/{symbol}` (Priority)

Replace `submit_order(MarketOrderRequest(SELL))` with Alpaca's native close endpoint:
- `close_position_emergency()` → `self.alpaca.close_position(symbol_or_id, percentage=100)`
- `scale_out_position()` → `self.alpaca.close_position(symbol_or_id, percentage=50)` 

Benefits:
- Uses Alpaca's actual qty (eliminates Firestore drift risk from #297)
- `percentage` param makes fractional scale-outs precise
- Atomic and idempotent
- Works for both equity and crypto

### 2. Evaluate `current_stop_loss` as a floor in `check_exits()` (Enhancement)

Add `current_stop_loss` from the linked Position as a secondary exit trigger in `check_exits()`. This would require passing Position data into the signal evaluation loop, or cross-referencing in the TP automation block.

```python
# Proposed: In main.py TP automation block, after check_exits() returns
# Add a defensive check: if current_close < position.current_stop_loss → emergency close
if pos and pos.current_stop_loss and current_close < pos.current_stop_loss:
    # Stop hit — close immediately regardless of signal status
    close_position_emergency(pos)
```

### 3. Consider `stop_limit` orders for between-run protection (Future)

Alpaca supports `stop_limit` for crypto (the only stop-type order available). This could provide broker-side stop protection between runs. However, `stop_limit` has **gap risk** — if price crashes through the limit price, the order won't fill. A possible mitigation is using a wide limit spread (e.g., stop at $100, limit at $90).

This would require:
- Submitting a `StopLimitOrderRequest` after entry
- Managing it like equity SL orders (replace on trail updates)
- Handling the case where the stop_limit fills between runs

## Technical Details / Tasks

- [ ] Replace `submit_order(MarketOrderRequest(SELL))` with `self.alpaca.close_position()` in `close_position_emergency()`
- [ ] Replace `submit_order(MarketOrderRequest(SELL))` with `self.alpaca.close_position()` in `scale_out_position()`
- [ ] Verify `close_position()` works for both equity and crypto symbols
- [ ] Add qty reconciliation: before close, fetch Alpaca's actual qty and compare with Firestore
- [ ] Evaluate whether `current_stop_loss` should be checked as exit trigger in the TP automation block
- [ ] Document Alpaca crypto order constraints in strategy docs
- [ ] Investigate `stop_limit` as broker-side SL for between-run protection (spike)

## Acceptance Criteria

- [ ] `close_position_emergency()` uses `DELETE /v2/positions/{symbol}` instead of `submit_order(SELL)`
- [ ] `scale_out_position()` uses `DELETE /v2/positions/{symbol}?percentage=50` or `qty=X`
- [ ] Position close operations use Alpaca's actual qty, not Firestore's potentially stale qty
- [ ] Existing TP automation flow (TP1 scale-out, TP2 scale-out, TP3 close) continues to work
- [ ] Strategy docs updated with Alpaca crypto order constraints appendix

## Appendix A: Alpaca Crypto Order Constraints

This section documents the Alpaca API constraints for crypto trading, which explain why our system uses the current architecture.

### Why No Bracket Orders for Crypto
Alpaca explicitly does not support bracket/OTOCO orders for crypto assets (error code `42210000`). This is why `_execute_crypto_signal()` uses a simple `MarketOrderRequest` while `_execute_bracket_order()` is reserved for equities.

### Why No Plain Stop Orders for Crypto
Alpaca supports `stop_limit` but NOT plain `stop` orders for crypto. A `stop_limit` requires both a `stop_price` (trigger) and a `limit_price` (maximum acceptable fill), introducing gap risk. Our decision to track stops in Firestore and evaluate them per-run via `check_exits()` was the correct trade-off.

### Available Exit Mechanisms for Crypto

| Method | Mechanism | When Used |
|---|---|---|
| `market` sell order | `POST /v2/orders` type=`market` | Current: emergency close, scale-outs |
| Close position | `DELETE /v2/positions/{symbol}` | **Proposed**: cleaner alternative to market sell |
| `stop_limit` order | `POST /v2/orders` type=`stop_limit` | Future: between-run SL protection |

### Order Lifecycle for Crypto (Current vs. Proposed)

```
ENTRY:
  POST /v2/orders (market, GTC) ← Current and correct

TP1 SCALE-OUT:
  Current:  POST /v2/orders (market SELL, qty=50%)
  Proposed: DELETE /v2/positions/{symbol}?percentage=50

TP2 SCALE-OUT:
  Current:  POST /v2/orders (market SELL, qty=50% of remaining)
  Proposed: DELETE /v2/positions/{symbol}?percentage=50

TP3 / INVALIDATION CLOSE:
  Current:  POST /v2/orders (market SELL, qty=all)
  Proposed: DELETE /v2/positions/{symbol}?percentage=100

BETWEEN-RUN SL (FUTURE):
  POST /v2/orders (stop_limit, stop_price=SL, limit_price=SL-5%)
```

### References
- [Crypto Orders API](https://docs.alpaca.markets/docs/crypto-orders)
- [Crypto Spot Trading](https://docs.alpaca.markets/docs/crypto-trading)
- [Close Position API](https://docs.alpaca.markets/reference/deleteopenposition-1)
- [Create Order API](https://docs.alpaca.markets/reference/postorder)

### PR DESCRIPTION
**Pull request by lagarcess**
This PR migrates the crypto position exit logic from manual market sell orders to Alpaca's native position liquidation API (`DELETE /v2/positions/{symbol}`). This improvement eliminates risks associated with quantity drift and ensures precise fractional scale-outs.

Additionally, the PR hardens the position lifecycle by evaluating the `current_stop_loss` field (tracked in Firestore) as a defensive floor in the main execution loop, providing protection even if candlestick indicators haven't yet triggered an exit.

Key Changes:
- `ExecutionEngine.close_position_emergency`: Now uses `alpaca.close_position(percentage="100")`.
- `ExecutionEngine.scale_out_position`: Now uses `alpaca.close_position(percentage="50")` (or other percentage).
- `main.py`: Added a defensive check for `current_stop_loss` before standard exit evaluation.
- `docs/strategy/position-management.md`: Added documentation for Alpaca crypto constraints.
- Updated unit tests in `tests/engine/test_execution.py` and `tests/engine/test_exit_price_capture.py` to verify the new API integration.

Fixes #301

---
*PR created automatically by Jules for task [7848239148622067016](https://jules.google.com/task/7848239148622067016) started by @lagarcess*

### CODE REVIEW

**Review by lagarcess — COMMENTED** [review]
lgtm
